src/main/java/com/google/cloud/tools/appengine/v2/deploy/DeployExample.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.appengine.v2.deploy;
+
+import com.google.cloud.tools.appengine.api.deploy.DefaultDeployConfiguration;
+import com.google.cloud.tools.appengine.cloudsdk.PathResolver;
+import com.google.cloud.tools.appengine.v2.AppEngine;
+import com.google.cloud.tools.appengine.v2.AppEngineRequestCallback;
+import com.google.cloud.tools.appengine.v2.AppEngineRequestFactory;
+import com.google.cloud.tools.appengine.v2.internal.process.io.PrintStreamOutputHandler;
+
+import java.io.File;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+public class DeployExample {
+
+  public static void main(String[] args) {
+
+    // perhaps we should be using builders for execute configurations
+    DefaultDeployConfiguration config = new DefaultDeployConfiguration();
+    config.setDeployables(Collections.singletonList(new File("/tmp/test.sh")));
+
+    // the current implementation doesn't have a good hook in to the autodetection of the
+    // cloud sdk.
+    AppEngineRequestFactory requestFactory = AppEngine.newRequestFactory()
+        .cloudSdk(Paths.get("/path/to/cloudsk"))
+        // or explicitly tell it to look for it
+        //.cloudSdk(new PathResolver().getCloudSdkPath()) //<<-- path resolver would go into the public API
+        .build();
+
+    // do the execute,
+    // the implementation of Deployment request doesn't allow modifiying the DeploymentRequest
+    // after execute() is called by enforcing it with illegal state exceptions? Is that necessary?
+    // Does the builder style imply that would be the case? Who knows...
+    Future<DeployResult> deployFuture = requestFactory.newDeploymentRequest(config)
+        .outputHandler(new PrintStreamOutputHandler(System.out))
+        .callback(new AppEngineRequestCallback<DeployResult>() {
+          @Override
+          public void onSuccess(DeployResult result) {
```

Comment:
How do you cancel the process if you have the callback only?

On Wed, Aug 10, 2016, 17:30 Appu Goundan <notifications@github.com> wrote:

> In
> src/main/java/com/google/cloud/tools/appengine/v2/deploy/DeployExample.java
> <https://github.com/GoogleCloudPlatform/appengine-plugins-core/pull/182#discussion_r74334678>
> :
>
> > +    // cloud sdk.
> > +    AppEngineRequestFactory requestFactory = AppEngine.newRequestFactory()
> > +        .cloudSdk(Paths.get("/path/to/cloudsk"))
> > +        // or explicitly tell it to look for it
> > +        //.cloudSdk(new PathResolver().getCloudSdkPath()) //<<-- path resolver would go into the public API
> > +        .build();
> > +
> > +    // do the execute,
> > +    // the implementation of Deployment request doesn't allow modifiying the DeploymentRequest
> > +    // after execute() is called by enforcing it with illegal state exceptions? Is that necessary?
> > +    // Does the builder style imply that would be the case? Who knows...
> > +    Future<DeployResult> deployFuture = requestFactory.newDeploymentRequest(config)
> > +        .outputHandler(new PrintStreamOutputHandler(System.out))
> > +        .callback(new AppEngineRequestCallback<DeployResult>() {
> > +          @Override
> > +          public void onSuccess(DeployResult result) {
>
> I think ListenableFuture requires you to have access to the executor
> underneath, which is weird, so it would have to be some kind of
> passthrough, but I have some doubts about how that'll work especially if
> registration of the callback happens after the process is finished (do race
> conditions come into effect here? or do I misunderstand how this would work)
>
> I like the Future style because it works well for the build tools, we
> could provide a basic wrapper for users wanting to register callbacks off
> their future (which is kind of the opposite of what Brian is proposing)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/appengine-plugins-core/pull/182/files/348d9bbb3d471fa2d917dce083d3b5ac42b0af7a..315f3fb531cd9561e627ead0ec1091b95ec7fbb7#r74334678>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADOO_vsTXC7sRIPoiwTp8YNmUo_T-eLXks5qekLggaJpZM4JYOIg>
> .
>